### ISSUE TITLE
In the context of a py 2 resolve, pip chooses a py 3 only dep and fails

### ISSUE THREAD

**Issue by stuhood** [pantsbuild/pex#947]
In the context of a build for a python 2 target on OSX, a command similar to the following is generated by PEX `2.1.6`:
```
~/src/venvs/pex/bin/pip -vvvv \
  --isolated \
  -q \
  --cache-dir \
  /Users/example/repo/.pants.d/python-setup/resolved_requirements/CPython-2.7.16 \
  download \
  --dest \
  blah \
  --no-index \
  --find-links \
  https://example.com/home/third_party/source/python/bootstrap/ \
  --find-links \
  https://example.com/home/third_party/source/python/wheels/ \
  --find-links \
  https://example.com/home/third_party/source/python/deprecated/ \
  --pre \
  mock==2.0.0 \
  pytest==4.1.1 \
  attrs==18.2.0 \
  Pillow==5.4.1 \
  'futures==3.2.0; python_version < "3"' \
  notebook==5.0.0 \
  jupyter_console==5.2.0 \
  ipython==5.8.0 \
  ipykernel==4.10.0 \
  appnope==0.1.0 \
  matplotlib==2.2.2 \
  numpy==1.16.4 \
  scipy==1.1.0 \
  tensorflow==1.14.0
```
...which fails with:
```
Package 'Markdown' requires a different Python: 2.7.16 not in '>=3.5'
```

After the error, the following indicates that one library has a dep on Markdown:
```
for whl in *.whl; do
  md=$(zipinfo -1 "$whl" | grep METADATA)
  unzip -c "$whl" "$md" | grep -i 'Requires-Dist.*Markdown' | awk "{print \"$whl:\t\"\$0}"
done
```
...
```
tensorboard-1.14.0-py2-none-any.whl:	Requires-Dist: markdown (>=2.6.8)
```

Although there are multiple copies of Markdown available (`2.1.1`, `2.6.2`, `2.6.11`, `3.0.1`), pip chooses to resolve `3.2`, which is not compatible with Python 2.

----

It's entirely possible that pip is doing something dumb here (although to be honest, I'm not seeing where PEX declares the python version it is resolving for, and that could be an issue?), but I think that as with the discussion from https://pantsbuild.slack.com/archives/C087V4P1T/p1584492639032900 , the solution might be to "ignore errors", and then include the problematic constraint in the error message.

**Comment by jsirois** [comment]
> It's entirely possible that pip is doing something dumb here (although to be honest, I'm not seeing where PEX declares the python version it is resolving for, and that could be an issue?)

For local resolves (no --platform in play), the correct local interpreter is used. In your example command that's whatever the interpreter that built the ven is (`~/src/venvs/pex/bin/python --version`).
The `--cache-dir` argument implies this is a use of Pex from Pants and that Pants thinks its running Pex with python 2.7.16. Can you confirm or deny these bits?

As for the failure, I can't repro, but I also don't have access to your `--find-links` and so details of wheels and wheel metadata found there could become critical:
```
$ curl -sSL https://github.com/pantsbuild/pex/releases/download/v2.1.6/pex > pex-2.1.6.pex && chmod +x pex-2.1.6.pex
$ ./pex-2.1.6.pex --python=python2.7 \
> mock==2.0.0 \
> pytest==4.1.1 \
> attrs==18.2.0 \
> Pillow==5.4.1 \
> 'futures==3.2.0; python_version < "3"' \
> notebook==5.0.0 \
> jupyter_console==5.2.0 \
> ipython==5.8.0 \
> ipykernel==4.10.0 \
> appnope==0.1.0 \
> matplotlib==2.2.2 \
> numpy==1.16.4 \
> scipy==1.1.0 \
> tensorflow==1.14.0 \
> -o issues-947.pex
$ ./issues-947.pex 
Python 2.7.17 (default, Mar 21 2020, 00:47:07) 
[GCC 9.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import tensorflow
>>> tensorflow.__file__
'/home/jsirois/.pex/installed_wheels/b35dc048b6d182e12b1690eeaa59a97f64e7a5ae/tensorflow-1.14.0-cp27-cp27mu-manylinux1_x86_64.whl/tensorflow/__init__.py'
>>> import markdown
>>> markdown.__file__
'/home/jsirois/.pex/installed_wheels/b80598369cacd1f28b9746dd5469f2573e545178/Markdown-3.1.1-py2.py3-none-any.whl/markdown/__init__.py'
>>> 
```

**Comment by jsirois** [comment]
@stuhood is there any way you can provide the relevant contents of the `--find-links` repos? For example, against my local pip cache:
```
find ~/.cache/pip/wheels/ -type f -name "*.whl" | while read wheel; do echo "> $wheel dependencies:" && unzip -qc $wheel '*/METADATA' | grep -E "^Requires-(Dist|Python): " | sort -u; done
...
> /home/jsirois/.cache/pip/wheels/12/92/12/23f91f6e8af1d180a113403f1643b25a0e2eedd6a195938f31/pex-2.1.6-cp36-none-any.whl dependencies:
Requires-Dist: certifi (==2016.9.26) ; extra == 'certs'
Requires-Dist: pyparsing (>=2.0.2)
Requires-Dist: pytest (>=3.0.0) ; extra == 'test'
Requires-Dist: pytest-cov ; extra == 'test'
Requires-Dist: six
Requires-Dist: subprocess32>=3.2.7 ; extra == "subprocess" and ( python_version<'3')
Requires-Dist: wincertstore (==0.2) ; (sys_platform == "win32") and extra == 'ssl'
Requires-Python: >=2.6, !=3.0.*, !=3.1.*, !=3.2.*
Requires-Python: >=2.7, !=3.0.*, !=3.1.*, !=3.2.*
Requires-Python: >=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*
Requires-Python: >=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*
Requires-Python: >=2.7,<3.9,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*
Requires-Python: !=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,>=2.7
> /home/jsirois/.cache/pip/wheels/12/92/12/23f91f6e8af1d180a113403f1643b25a0e2eedd6a195938f31/pex-2.1.6-cp37-none-any.whl dependencies:
Requires-Dist: certifi (==2016.9.26) ; extra == 'certs'
Requires-Dist: pyparsing (>=2.0.2)
Requires-Dist: pytest (>=3.0.0) ; extra == 'test'
Requires-Dist: pytest-cov ; extra == 'test'
Requires-Dist: six
Requires-Dist: subprocess32>=3.2.7 ; extra == "subprocess" and ( python_version<'3')
Requires-Dist: wincertstore (==0.2) ; (sys_platform == "win32") and extra == 'ssl'
Requires-Python: >=2.6, !=3.0.*, !=3.1.*, !=3.2.*
Requires-Python: >=2.7, !=3.0.*, !=3.1.*, !=3.2.*
Requires-Python: >=2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*
Requires-Python: >=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*
Requires-Python: >=2.7,<3.9,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*
Requires-Python: !=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,>=2.7
> /home/jsirois/.cache/pip/wheels/ae/79/41/345e8e433cfb890c43385170c5da68cbffeba67210c5bd111c/twitter.common.options-0.3.11-py3-none-any.whl dependencies:
> /home/jsirois/.cache/pip/wheels/ae/f3/c6/ffd1b0df0828cf685b0c863655f166b426cf1a3f62578206f1/pywatchman-1.4.1-cp38-cp38-linux_x86_64.whl dependencies:
> /home/jsirois/.cache/pip/wheels/ad/fc/61/1c76ae24b519301d5707177ca5492d0545aecf613c02287f1b/twitter.common.lang-0.3.11-py3-none-any.whl dependencies:
> /home/jsirois/.cache/pip/wheels/6d/f5/c5/6b26dd61ae703ab9ae4da42794fc3d574cdf5c7353e57b70f7/twitter.common.log-0.3.11-py3-none-any.whl dependencies:
Requires-Dist: twitter.common.dirutil (==0.3.11)
Requires-Dist: twitter.common.options (==0.3.11)
```

**Comment by cosmicexplorer** [comment]
It's possible that #1033 could make this situation less mysterious or remove it altogether, by removing any implicit dependency on the interpreter versions available on the local disk.

**Comment by jsirois** [comment]
Since this has gone along time without extra information and:

> For local resolves (no --platform in play), the correct local interpreter is used. 

This assertion has been proved false specifically for macOS venvs that involve macOS framework build Pythons as documented in #1009.

I'm going to close as likely fixed by #1049 and #1065.
